### PR TITLE
[Java] Support addition of gradient operations in a graph

### PR DESCRIPTION
**Pull request by karllessard**
This calls the C-api `TF_AddGradients` method through a new JNI binding for adding gradient nodes to a graph. It also includes an `AddGradients` wrapper for invoking this operation smoothly while building a graph using the new Java Ops API.

### CODE REVIEW

**Review by asimshankar — CHANGES_REQUESTED** [review]
Thanks a bunch!

File: tensorflow/java/src/main/java/org/tensorflow/Graph.java [hunk 1]
```diff
@@ -143,6 +143,68 @@ public void importGraphDef(byte[] graphDef, String prefix) throws IllegalArgumen
...
+   * <p>
+   * If {@code dx} is null, the implementation will use dx of {@code OnesLike} for all
+   * shapes in {@code y}.
+   * 
+   * @param y
+   * @param x
+   * @param dx
+   * @return the partial derivatives {@code dy} with the size of {@code x}
+   */
+  public Output<?>[] addGradients(Output<?>[] y, Output<?>[] x, Output<?>[] dx) {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow/java/src/main/java/org/tensorflow/Graph.java]: Sounds very good to me. Would it matter if we do that modification only in the `Op` class? I kind of see that this kind of interface optimization is the responsibility of the Ops API layer while the core classes focuses more on the implementation details.

[the remark this answers is not in the document]
Comment (asimshankar) [reply tensorflow/java/src/main/java/org/tensorflow/Graph.java]: I don't feel terribly strongly, but have a mild preference to include the simple override in the`Graph` class as well.

Is the second `Op` you mentioned basically the equivalent of `Optimizer.minimize()` in Python? We can talk about that in a follow up, but as you suggested, not do that in this PR.

File: tensorflow/java/src/main/native/graph_jni.cc [hunk 2]
```diff
@@ -130,3 +132,55 @@ Java_org_tensorflow_Graph_toGraphDef(JNIEnv* env, jclass clazz, jlong handle) {
...
+  if (!throwExceptionIfNotOK(env, status)) {
+    TF_DeleteStatus(status);
+    return nullptr;
+  }
+  TF_DeleteStatus(status);
+
+  // returned array contains both op handles and output indices, in pair
+  jlongArray dy_handles_and_indices = env->NewLongArray(nx << 1);
+  jlong* dy_elems = env->GetLongArrayElements(dy_handles_and_indices, nullptr);
+  for (int i = 0, j = nx; i < nx; ++i, ++j) {
```
Comment (asimshankar) [inline tensorflow/java/src/main/native/graph_jni.cc]: In practice this probably won't matter at all since `dy_elems` will be a pretty small array, but did you consider encoding it as `[handle0, index0, handle1, index1, ...]` instead of `[handle0, handle1, ..., index0, index1, ...]`? The former would be more cache friendly :)

File: tensorflow/java/src/main/java/org/tensorflow/op/training/Gradients.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/op/training/Gradients.java]: Why `training/`? Seems like `gradients` could be a top level operation too? One can use gradients for things other than training :)

**Review by asimshankar — CHANGES_REQUESTED** [review]
Thanks for the update, apologies for the delay. A few more minor things, otherwise looks great!

File: tensorflow/java/src/main/java/org/tensorflow/Graph.java [hunk 4]
```diff
@@ -143,6 +143,82 @@ public void importGraphDef(byte[] graphDef, String prefix) throws IllegalArgumen
...
+   * <p>
+   * If {@code dx} is null, the implementation will use dx of {@link org.tensorflow.op.core.OnesLike OnesLike} for all
+   * shapes in {@code y}.
+   * 
+   * @param y output of the function to derive
+   * @param x inputs of the function for which partial derivatives are computed
+   * @param dx if not null, the partial derivatives of some loss function {@code L} w.r.t. {@code y}
+   * @return the partial derivatives {@code dy} with the size of {@code x}
+   */
+  public Output<?>[] addGradients(Output<?>[] y, Output<?>[] x, Output<?>[] dx) {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow/java/src/main/java/org/tensorflow/Graph.java]: I wasn't sure about this, you probably know better than me: is it possible to have a graph with variables of different types? If it is guaranteed that could never happen, I'll gladly remove those wildcards to enforce type-safety.

File: tensorflow/java/src/main/native/graph_jni.cc (hunk 2 above)
Comment (karllessard) [reply tensorflow/java/src/main/native/graph_jni.cc]: Ironically, I was using the format you proposed in my previous commit. I switched to the new one lately because I found it more elegant to iterate using two iterators at the same time and also because I thought we might find some optimization later to simply split the array in two instead of copying its elements one by one. 

I don't remind reverting to the previous version if you want to.

File: tensorflow/java/src/main/java/org/tensorflow/op/training/Gradients.java (hunk 3 above)
Comment (karllessard) [reply tensorflow/java/src/main/java/org/tensorflow/op/training/Gradients.java]: No problem. Right now, the `Op` classification is still obscure to me, I'm taking some guesses here and there but we will probably need to settle this at some point?

File: tensorflow/java/src/main/java/org/tensorflow/Graph.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (asimshankar) [reply tensorflow/java/src/main/java/org/tensorflow/Graph.java]: Hmm...it actually is, since one could have say `tf.cast`s on the path between `x` and `y`, something like:

```python
import tensorflow as tf

x = tf.placeholder(tf.float64)
y = tf.square(tf.cast(x, tf.float32))
dy = tf.gradients(y, x)[0]
print(x.dtype, y.dtype, dy.dtype)
```

So apologies, ignore my comment :)

File: tensorflow/java/src/main/native/graph_jni.cc (hunk 2 above)
Comment (asimshankar) [reply tensorflow/java/src/main/native/graph_jni.cc]: It's an internal detail, so we can easily switch it around as we go along. I don't feel very strongly, but am slightly partial to the previous version. Your call.

File: tensorflow/java/src/main/java/org/tensorflow/op/training/Gradients.java (hunk 3 above)
Comment (asimshankar) [reply tensorflow/java/src/main/java/org/tensorflow/op/training/Gradients.java]: BTW, the Python code was recently updated to generate symbols in finer grained namespaces (https://github.com/tensorflow/tensorflow/commit/c1ff1164e30186d847f7d4f9e9ce5d40936a2c1c#diff-d75a97d0f69d6f87ab79be6e2423d87b). So perhaps we could use the same for Java (without the baggage of backward compatibility. For example, `Acos` can be just in the `math` namespace and not in the top level one).

If you're enthusiastic, I'd be more than happy to see a PR adding Java API defs :)

File: tensorflow/java/src/test/java/org/tensorflow/TestUtil.java [hunk 5]
```diff
@@ -36,7 +36,7 @@
         .<T>output(0);
   }
 
-  public static Output<?> addN(Graph g, Output<?>... inputs) {
+  public static <T> Output<T> addN(Graph g, Output<?>... inputs) {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow/java/src/test/java/org/tensorflow/TestUtil.java]: Unfortunately, the compiler complains with a warning when using a parameterized type for a varargs parameter.

**Review by asimshankar — APPROVED** [review]

**Comment by qlzh727** [comment]
Windows build failure seems to be a flaky. Submitting now.
